docs/features/role-based-access-control/high-level-design.md

```diff
@@ -0,0 +1,81 @@
+# Role base access control - High-level design
+
+This document aims at providing an overview of the architecture of Role-based Access Control (RBAC from here on) for
+the Marxan platform.
+
+## Roles
+
+Platform administrators should be identified in a database table dedicated for the purpose. Adding users to the admins
+DB table should only be done by other admin users.
+
+The remaining roles supported by the platform (`Owner`, `Contributor` and `Viewer`) will be stored in the `roles` 
+database in the `marxan-api` database. These roles should cover the roles defined in the project requirements:
+
+* "Platform admin" would be achieved with the implementation of a dedicated DB table for platform admins;
+* "Project owner" -> role Owner for resource Project;
+* "Project contributor" -> role Contributor for resource Project;
+* "Project reviewer (read-only)" -> role Viewer for resource Project;
+* "Project solution-viewer" -> role Viewer for resource "Solutions";
+
+Roles should have numerical values associated to them, to identify the hierarchical order between them. For instance:
+
+* `Owner` -> 1000
+* `Contributor` -> 500
+* `Viewer` -> 250
+
+Roles with fewer privileges should have lower numerical values than roles with more privileges. Using numerical values
+for roles allows for easily comparing if a given user has certain privileges (e.g. checking if a user has the role 
+`Contributor` can be easily done: `role >= Contributor (500)`).
+
+There is also an implicit relationship between platform actions and the supported roles. By default, it is assumed that:
+  * a user with role `Viewer` can read the information for the given resource (which usually includes calls to 
+`GET resource` and `GET resource/:id` endpoints).
+  * user with role `Contributor` can perform all the actions a `Viewer` can, as well as edition actions for the given
+resource (which usually includes calls to `PATCH/PUT resource/:id` endpoints).
+  * user with `Owner` role can perform all the actions a `Contributor` can, as well as deleting actions for the given
+resource (which usually includes calls to `DELETE resource/:id` endpoints).
+
+**However, it is up to the functions implementing each feature (modules, controllers, services or any other kind of 
+object in the context of the API) to correctly check permissions for the action being performed.** This also applies
+to custom actions that might be subject to RBAC: for instance, if publishing a project is a custom action that should be
+subject to RBAC, then it is up to the controller action, service method, or wherever we need to do this, to check that 
+the user performing the request has the required role (e.g. `Owner`) to perform that action.
+
+Currently, these are the roles required to perform the following actions in the platform:
```

Comment:
This solves most of the doubts I raised in my review of brief.md 😅

However:

- should we limit permissions of "higher" roles on parent resources (e.g. Organization Owner) to read-only ones for children resources?

Obviously this is not meant as a privilege escalation measure (at least Owners could in any case update roles), but mostly as a least-privilege approach to protect against accidental changes that an Organization Owner, for example, may end up making. For roles that cannot manage roles on a resource this can also provide some protection against privilege escalation.

- should we mark the 'implicit' permissions as such in this table? I am ambivalent about this - in a way, any permission can be seen as just that, a permission, no matter whether the role has "Organization" in their name but a permission given to this role applies to a scenario; on the other hand, if it helps to reason about access levels, it may be useful

Reply:
Also, I think we have a chance *right now* to model many-to-many links between at least organizations and projects, which is something that was discussed much earlier on as a wish (for example, project A belongs both to organization M and organization N), with related composition of permissions (if user U has contributor role on organization M and no role in organization N, they can see the project) - later on this may be quite messy to disentangle. Happy to go with one-to-many links in any case if m-to-m adds too much right now, but let's be explicit about this maybe.

Reply:
As discussed during today's daily meeting, I'll remove the "hierarchical" bits from this implementation, which will also imply we remove permissions granted implicitly.

This gives us a more flexible approach in the long term, and we still have the possibility of programatically granting permissions with events. Plus, it doesn't constrain us to having to model many-to-many links right now, we can do it later down the road without compromising the existing permissions structure. 

This also addresses your point of "marking implicit permissions in this table" - without implicit permissions in place, this is not a concern because all permissions will have to be explicitly set in the permissions table :+1: